JOSS Review: Summary of issues building ASGarD with HDF5 on M1

## Describe the bug
This is part of my JOSS review for openjournals/joss-reviews#4378.

The build process on M1 using conda package management has been pretty straightforward for a basic installation, except when attempting to use a pre-installed HDF5, or allowing ASGarD to build it itself. Usage of HighFive and HDF5 with ASGarD was not clear in my reading of the README or Wiki documentation, so I had to work a bit blind. I've outlined issues experienced with each case below that might need to be addressed via documentation or CMake adjustments:

1. Using `-DASGARD_HDF5_PATH` to point to a pre-installed HDF5 located at `~/mambaforge3/envs/asgard/`. Specifics of the packages installed are covered below under the System heading. 
    
    - Starting from [CMakeLists.txt](https://github.com/project-asgard/asgard/blob/4046c0a264ca1d156be1ecf5f8a30d290a0a7720/CMakeLists.txt), the `ASGARD_HDF5_PATH` variable was initially used to point ASGarD to the conda prefix for HDF5. Using 
       ```
       cmake .. -DCMAKE_CXX_COMPILER=mpicxx -DASGARD_USE_MPI=YES -DASGARD_HDF5_PATH=~/mambaforge3/envs/asgard/
       ``` 
       led to no hint of HDF5 in the CMake output. It appeared as though this wasn't the correct/primary variable to start with, so another attempt was made adding on a reference to HighFive.
    - Adding `-DASGARD_IO_HIGHFIVE=ON` to the prior CMake command led to the following console output:
        ```
        -- using external hdf5 found at ~/mambaforge3/envs/asgard/
        -- using contrib HighFive at /Users/cticenhour/projects/asgard/contrib/highfive
        ```
        which is what I expected to see given the references to HighFive and HDF5 in the documentation. Sure enough, `io-test` was added to the test listing after build. If both of these variables are required, should `ASGARD_IO_HIGHFIVE` be set to ON whenever `ADGARD_HDF5_PATH` is set using CMake dependent options? 

2. Letting ASGarD build its own HDF5

    - Again, looking at `CMakeLists.txt`, it appears as though `ASGARD_BUILD_HDF5` is the primary variable of interest here. Removing the installed HDF5 from earlier, and using:
        ```
        cmake .. -DCMAKE_CXX_COMPILER=mpicxx -DASGARD_USE_MPI=YES -DASGARD_BUILD_HDF5=ON
        ```
        we again don't have any hint of HDF5 in the cmake output. Adding `-DASGARD_IO_HIGHFIVE=ON` yields the following successful console output:
        ```
        -- using external hdf5 found at HDF5_hdf5_LIBRARY-NOTFOUND
        -- libhdf5 not found. We'll build it from source.
        -- using contrib HighFive at /Users/cticenhour/projects/asgard/contrib/highfive
        ```
        Without looking at `contrib/io.cmake` and purely at the console, it appears as though `ASGARD_BUILD_HDF5` does nothing, so it should likely also be tied to `ASGARD_IO_HIGHFIVE` as a dependency whenever it is used.
    
    - When the build begins, HDF5 1.10.5 is downloaded and the build begins. This fails with the error: 
        ```
        checking build system type... Invalid configuration `arm64-apple-darwin20.0.0': machine `arm64-apple' not recognized
        configure: error: /bin/sh bin/config.sub arm64-apple-darwin20.0.0 failed
        ``` 
        I've experienced this before when working with HDF5 in the MOOSE framework, in idaholab/moose#20736, which I had to fix by adding a pre-configure step to the build for M1 platforms. I made a patch below relevant to this case:
        ```
        diff --git a/contrib/io.cmake b/contrib/io.cmake
        index 22f5c3c4..91ad9c7a 100644
        --- a/contrib/io.cmake
        +++ b/contrib/io.cmake
        @@ -44,7 +44,8 @@ function (get_hdf5)
                 PREFIX contrib/hdf5
                 URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2
                 DOWNLOAD_NO_PROGRESS 1
        -        CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
        +        CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/autogen.sh
        +        COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
                 BUILD_COMMAND make
                 BUILD_IN_SOURCE 1
                 INSTALL_COMMAND make install
        ```
        which allows HDF5 to reconfigure itself using `automake` and `libtool`. For Darwin ARM64 platforms using homebrew, this change should be adequate as `autogen.sh` looks for the location of `glibtool` specifically (which homebrew provides), but for conda an extra step needs to be taken. One either needs to specify the location of `libtool` using the `HDF5_LIBTOOL` variable, or create a link to `libtool` named `glibtool` for the script to pick up. Either way, HDF5 should then build successfully and pass the added `io-test`. 

## To Reproduce

CMake specifics for each case are outlined above, but common attributes for the ASGarD code as cloned on my system for each are outlined below: 
1. Using the `develop` branch, commit 4046c0a264ca1d156be1ecf5f8a30d290a0a7720
2. Make command: `make -j10`
4. Test invocation: `ctest -j10`

## Expected behavior
All required tests for a configuration including HDF5 (with MPI in this case) should pass successfully.

## System
 - `MacBookPro18,2`: Apple M1 Max, 64GB RAM, macOS Monterey 12.5, Command Line Tools 13.4.0.0.1.1651278267
 - Using conda package management (using packages from `conda-forge`) with the following install command: 
    ```
    conda install openmpi=4.1.4 openmpi-mpicc=4.1.4 openmpi-mpicxx=4.1.4 cmake=3.23.3 hdf5=1.12.2=mpi_openmpi_hfe3cbcd_0
    ```
    - This package configuration uses Clang 13.0.1
    - The HDF5 package was removed during HDF5 build testing. 
 - When allowing ASGarD to build its own HDF5, more packages were needed for a pre-configuration step:
     ```
     conda install automake=2.69 libtool=2.4.6
     ```
     - This led to:
         - Autoconf 2.69
         - Automake 1.16.5
         - Libtool 2.4.6
         - M4 1.4.18

## Additional context
I still plan to test this workflow on an Intel Mac and an AMD based Linux environment with a similar configuration (will note differences here with an edit).

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -51,6 +51,15 @@ register_project (Catch2
                   v3.0.1
 )
 
+#Mark CATCH variables as advanced.
+get_cmake_property(_variableNames VARIABLES)
+foreach (_variableName ${_variableNames})
+    string(FIND "${_variableName}" "CATCH_" out)
+    if("${out}" EQUAL 0)
+        mark_as_advanced(${_variableName})
+    endif()
+endforeach()
+
 ###############################################################################
 ## Set up the compiler and general global build options
 ###############################################################################
@@ -79,10 +88,13 @@ endif ()
 string(APPEND CMAKE_CXX_FLAGS " -Wall -Wextra -Wpedantic")
 
 # set up possible commandline input variable defaults (override with -D)
+include(CMakeDependentOption)
 option (ASGARD_BUILD_TESTS "Build tests for asgard" ON)
 option (ASGARD_BUILD_OPENBLAS "Download and build our own OpenBLAS" OFF)
+option (ASGARD_IO_HIGHFIVE "Use the HighFive HDF5 header library for I/O" OFF)
 option (ASGARD_BUILD_HDF5 "Download and build our own HDF5/HighFive" OFF)
-option (ASGARD_HDF5_PATH "optional location of lib/ containing HDF5" "")
+option (ASGARD_HIGHFIVE_PATH "Optional location of existing HighFive library" "")
+option (ASGARD_HDF5_PATH "Optional location of lib/ containing HDF5" "")
 option (ASGARD_BUILD_PROFILE_DEPS "Download and build our own tools (e.g.  graphviz)" OFF)
 option (ASGARD_LAPACK_PATH "optional location of include/ and lib/ containing LAPACK" "")
 option (ASGARD_BLAS_PATH "optional location of include/ and lib/ containing BLAS" "")
@@ -93,12 +105,10 @@ option (ASGARD_PROFILE_GPERF_MEM "enable memory profiling support using gperftoo
 option (ASGARD_PROFILE_PERF "enable profiling support for using linux perf" "")
 option (ASGARD_PROFILE_VALGRIND "enable profiling support for using valgrind" "")
 option (ASGARD_GRAPHVIZ_PATH "optional location of bin/ containing dot executable" "")
-option (ASGARD_IO_HIGHFIVE "Use the HighFive HDF5 header library for I/O" OFF)
 option (ASGARD_USE_CUDA "Optional CUDA support for asgard" OFF)
 option (ASGARD_USE_OPENMP "Optional openMP support for asgard" ON)
 option (ASGARD_USE_MPI "Optional distributed computing support for asgard" OFF)
-include(CMakeDependentOption)
-cmake_dependent_option(ASGARD_USE_SCALAPACK "Use optional scalapack linear algebra library" OFF "ASGARD_USE_MPI" OFF)
+cmake_dependent_option (ASGARD_USE_SCALAPACK "Use optional scalapack linear algebra library" OFF "ASGARD_USE_MPI" OFF)
 option (ASGARD_USE_MATLAB "Optional MATLAB plotting support for asgard" OFF)
 option (ASGARD_USE_DOUBLE_PREC "enable or disable double-precision floating point" ON)
 option (ASGARD_USE_PCH "Enable precompiled header files." OFF)
@@ -117,6 +127,14 @@ endif()
 if (NOT ASGARD_LAPACK_PATH AND ASGARD_BLAS_PATH)
   set (ASGARD_LAPACK_PATH ${ASGARD_BLAS_PATH})
 endif()
+
+#Enable HighFive IO if any of the following variables are defined
+if(NOT ASGARD_IO_HIGHFIVE)
+  if(ASGARD_HDF5_PATH OR ASGARD_BUILD_HDF5 OR ASGARD_HIGHFIVE_PATH)
+    set(ASGARD_IO_HIGHFIVE ON CACHE BOOL "" FORCE)
+  endif()
+endif()
+
 ###############################################################################
 ## Pull in external support as needed
 ###############################################################################
@@ -271,7 +289,7 @@ target_link_libraries (asgard_obj
                        $<$<BOOL:${LINALG_FOUND}>:LINALG::LINALG>
                        kron
                        clara
-                       $<$<BOOL:${ASGARD_IO_HIGHFIVE}>:highfive>
+                       $<$<BOOL:${ASGARD_IO_HIGHFIVE}>:HighFive>
                        $<$<BOOL:${ASGARD_IO_HIGHFIVE}>:hdf5>
 )
 target_include_directories (asgard_obj

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -210,6 +210,10 @@ configure_file(
 
 add_library (asgard_obj)
 
+if (build_hdf5)
+    add_dependencies (asgard_obj hdf5-ext)
+endif ()
+
 target_compile_features (asgard_obj PUBLIC cxx_std_17)
 set (components
      adapt
@@ -285,7 +289,7 @@ target_link_libraries (asgard_obj
                        $<$<BOOL:${LINALG_FOUND}>:LINALG::LINALG>
                        kron
                        clara
-                       $<$<BOOL:${ASGARD_IO_HIGHFIVE}>:HighFive>
+                       $<$<BOOL:${ASGARD_IO_HIGHFIVE}>:highfive>
                        $<$<BOOL:${ASGARD_IO_HIGHFIVE}>:hdf5>
 )
 target_include_directories (asgard_obj

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -210,10 +210,6 @@ configure_file(
 
 add_library (asgard_obj)
 
-if (build_hdf5)
-    add_dependencies (asgard_obj hdf5-ext)
-endif ()
-
 target_compile_features (asgard_obj PUBLIC cxx_std_17)
 set (components
      adapt

--- a/README.md
+++ b/README.md
@@ -47,7 +47,7 @@ National Laboratory.
 
 * cuda
 * mpi
-* highfive/hdf5
+* HighFive/hdf5
 * MATLAB
 * ScaLAPACK
 
@@ -71,6 +71,13 @@ To see a list of available PDEs, run `./asgard --available_pdes`. The listed PDE
 To see the list of all runtime options, run `./asgard --help`.
 
 For specific platform build instructions, [see this wiki page.](https://github.com/project-asgard/asgard/wiki/platforms)
+
+## HighFive/HDF5 IO
+
+To enable HDF5 wavelet and realspace output, pass the `-DASGARD_IO_HIGHFIVE=ON`. 
+
+Pass `-DASGARD_BUILD_HDF5=TRUE` and ASGarD will download and build HDF5 and HighFive libraries. If HIGHFIVE and/or HDF5 are already 
+available, pass `-DASGARD_HIGHFIVE_PATH=<path to HighFive>` and/or `-DASGARD_HDF5_PATH=<path to HDF5>`.
 
 ## MATLAB Interface 
 

--- a/contrib/io.cmake
+++ b/contrib/io.cmake
@@ -40,13 +40,13 @@ function (get_hdf5)
 
       include (ExternalProject)
       if (DEFINED CMAKE_APPLE_SILICON_PROCESSOR and CMAKE_APPLE_SILICON_PROCESSOR STREQUAL "arm64")
-	# Get HDF5 to build on Apple silicon      
+        # Get HDF5 to build on Apple silicon
         ExternalProject_Add (hdf5-ext
           UPDATE_COMMAND ""
           PREFIX contrib/hdf5
           URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2
           DOWNLOAD_NO_PROGRESS 1
-	  CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/autogen.sh
+          CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/autogen.sh
           COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
           BUILD_COMMAND make
           BUILD_IN_SOURCE 1

--- a/contrib/io.cmake
+++ b/contrib/io.cmake
@@ -32,42 +32,49 @@ function (get_hdf5)
 
   # if cmake couldn't find other hdf5, or the user asked to build it
   if (NOT HDF5_FOUND)
-    set (hdf5_contrib_path ${CMAKE_SOURCE_DIR}/contrib/HDF5)
+    set (hdf5_contrib_path ${CMAKE_SOURCE_DIR}/contrib/hdf5)
     find_library (hdf5_search hdf5
       PATHS ${hdf5_contrib_path} PATH_SUFFIXES lib lib64 NO_DEFAULT_PATH)
     if (NOT hdf5_search)
       message (STATUS "libhdf5 not found. We'll build it from source.")
-      set( HDF5_TEST_CPP OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_EXAMPLES OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_FORTRAN OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_JAVA OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_PARALLEL OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_PASSTHROUGH_VOL OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_SERIAL OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_SWMR OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_TOOLS OFF CACHE BOOL "" FORCE)
-      set( HDF5_TEST_VFD OFF CACHE BOOL "" FORCE)
-      register_project(hdf5 HDF5 https://github.com/HDFGroup/hdf5 hdf5-1_12_2)
-      unset(CMAKE_RUNTIME_OUTPUT_DIRECTORY CACHE)
-      #Mark CATCH variables as advanced.
-      get_cmake_property(_variableNames VARIABLES)
-      foreach (_variableName ${_variableNames})
-	string(FIND "${_variableName}" "HDF5_" out)
-        if("${out}" EQUAL 0)
-          mark_as_advanced(${_variableName})
-        endif()
-      endforeach()
+
+      include (ExternalProject)
+      if (DEFINED CMAKE_APPLE_SILICON_PROCESSOR and CMAKE_APPLE_SILICON_PROCESSOR STREQUAL "arm64")
+	# Get HDF5 to build on Apple silicon      
+        ExternalProject_Add (hdf5-ext
+          UPDATE_COMMAND ""
+          PREFIX contrib/hdf5
+          URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2
+          DOWNLOAD_NO_PROGRESS 1
+	  CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/autogen.sh
+          COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
+          BUILD_COMMAND make
+          BUILD_IN_SOURCE 1
+          INSTALL_COMMAND make install
+        )
+      else()
+	ExternalProject_Add (hdf5-ext
+          UPDATE_COMMAND ""
+          PREFIX contrib/hdf5
+          URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2
+          DOWNLOAD_NO_PROGRESS 1
+          CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
+          BUILD_COMMAND make
+          BUILD_IN_SOURCE 1
+          INSTALL_COMMAND make install
+        )
+      endif()
       set (build_hdf5 TRUE PARENT_SCOPE)
-      set (hdf5_include ${HDF5_INCLUDE_DIRS})
-      set (hdf5_lib ${HDF5_LIBRARIES})
     else ()
       message (STATUS "using contrib HDF5 found at ${hdf5_search}")
     endif ()
     # either it was already here, or we just built it here
-    target_include_directories (hdf5 INTERFACE ${hdf5_include})
-    target_link_libraries (hdf5 INTERFACE ${hdf5_lib})
-
+    set (hdf5_include ${hdf5_contrib_path}/include)
+    set (hdf5_lib "-L${hdf5_contrib_path}/lib -Wl,-rpath,${hdf5_contrib_path}/lib/ -lhdf5")
   endif ()
+
+  target_include_directories (hdf5 INTERFACE ${hdf5_include})
+  target_link_libraries (hdf5 INTERFACE ${hdf5_lib})
 endfunction()
 
 ###############################################################################
@@ -83,14 +90,29 @@ if (ASGARD_IO_HIGHFIVE)
   get_hdf5()
 
   # now HighFive itself
-  if(ASGARD_HIGHFIVE_PATH AND NOT ASGARD_BUILD_HDF5)
-    find_library(Highfive PATHS ${hdf5_contrib_path} PATH_SUFFIXES lib lib64 NO_DEFAULT_PATH)
-  else()
-    set(HIGHFIVE_BUILD_DOCS OFF CACHE BOOL "" FORCE)
-    set(HIGHFIVE_EXAMPLES OFF CACHE BOOL "" FORCE)
-    set(HIGHFIVE_UNIT_TESTS OFF CACHE BOOL "" FORCE)
-    set(HIGHFIVE_USE_BOOST OFF CACHE BOOL "" FORCE)
-    register_project(highfive HIGHFIVE https://github.com/BlueBrain/HighFive v2.4.1)
-    message (STATUS "using contrib HighFive at ${HighFive_BINARY_DIR}")
-  endif()
+
+  set (highfive_PATH ${CMAKE_SOURCE_DIR}/contrib/highfive)
+  if (NOT EXISTS ${highfive_PATH}/include/highfive/H5Easy.hpp)
+    execute_process (COMMAND rm -rf ${highfive_PATH})
+    execute_process (COMMAND mkdir ${highfive_PATH})
+
+    message (STATUS "downloading HighFive from github")
+    execute_process (
+      COMMAND git clone https://github.com/BlueBrain/HighFive .
+      WORKING_DIRECTORY ${highfive_PATH}
+      RESULT_VARIABLE download
+      OUTPUT_QUIET
+      ERROR_QUIET
+      )
+    if (download)
+      message (FATAL_ERROR "could not download highfive")
+    endif ()
+  else ()
+    message (STATUS "using contrib HighFive at ${highfive_PATH}")
+  endif ()
+
+  add_library (highfive INTERFACE)
+  target_include_directories (highfive INTERFACE ${highfive_PATH}/include)
+  target_link_libraries (highfive INTERFACE hdf5)
+
 endif()

--- a/contrib/io.cmake
+++ b/contrib/io.cmake
@@ -44,8 +44,9 @@ function (get_hdf5)
         PREFIX contrib/hdf5
         URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2
         DOWNLOAD_NO_PROGRESS 1
-        CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
-        BUILD_COMMAND make
+        CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/autogen.sh
+        COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
+	BUILD_COMMAND make
         BUILD_IN_SOURCE 1
         INSTALL_COMMAND make install
       )
@@ -75,29 +76,14 @@ if (ASGARD_IO_HIGHFIVE)
   get_hdf5()
 
   # now HighFive itself
-
-  set (highfive_PATH ${CMAKE_SOURCE_DIR}/contrib/highfive)
-  if (NOT EXISTS ${highfive_PATH}/include/highfive/H5Easy.hpp)
-    execute_process (COMMAND rm -rf ${highfive_PATH})
-    execute_process (COMMAND mkdir ${highfive_PATH})
-
-    message (STATUS "downloading HighFive from github")
-    execute_process (
-      COMMAND git clone https://github.com/BlueBrain/HighFive .
-      WORKING_DIRECTORY ${highfive_PATH}
-      RESULT_VARIABLE download
-      OUTPUT_QUIET
-      ERROR_QUIET
-      )
-    if (download)
-      message (FATAL_ERROR "could not download highfive")
-    endif ()
-  else ()
-    message (STATUS "using contrib HighFive at ${highfive_PATH}")
-  endif ()
-
-  add_library (highfive INTERFACE)
-  target_include_directories (highfive INTERFACE ${highfive_PATH}/include)
-  target_link_libraries (highfive INTERFACE hdf5)
-
+  if(ASGARD_HIGHFIVE_PATH AND NOT ASGARD_BUILD_HDF5)
+    find_library(Highfive PATHS ${hdf5_contrib_path} PATH_SUFFIXES lib lib64 NO_DEFAULT_PATH)
+  else()
+    set(HIGHFIVE_BUILD_DOCS OFF CACHE BOOL "" FORCE)
+    set(HIGHFIVE_EXAMPLES OFF CACHE BOOL "" FORCE)
+    set(HIGHFIVE_UNIT_TESTS OFF CACHE BOOL "" FORCE)
+    set(HIGHFIVE_USE_BOOST OFF CACHE BOOL "" FORCE)
+    register_project(highfive HIGHFIVE https://github.com/BlueBrain/HighFive v2.4.1)
+    message (STATUS "using contrib HighFive at ${HighFive_BINARY_DIR}")
+  endif()
 endif()

--- a/contrib/io.cmake
+++ b/contrib/io.cmake
@@ -32,35 +32,42 @@ function (get_hdf5)
 
   # if cmake couldn't find other hdf5, or the user asked to build it
   if (NOT HDF5_FOUND)
-    set (hdf5_contrib_path ${CMAKE_SOURCE_DIR}/contrib/hdf5)
+    set (hdf5_contrib_path ${CMAKE_SOURCE_DIR}/contrib/HDF5)
     find_library (hdf5_search hdf5
       PATHS ${hdf5_contrib_path} PATH_SUFFIXES lib lib64 NO_DEFAULT_PATH)
     if (NOT hdf5_search)
       message (STATUS "libhdf5 not found. We'll build it from source.")
-
-      include (ExternalProject)
-      ExternalProject_Add (hdf5-ext
-        UPDATE_COMMAND ""
-        PREFIX contrib/hdf5
-        URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2
-        DOWNLOAD_NO_PROGRESS 1
-        CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/autogen.sh
-        COMMAND ${CMAKE_CURRENT_BINARY_DIR}/contrib/hdf5/src/hdf5-ext/configure --prefix=${hdf5_contrib_path}
-	BUILD_COMMAND make
-        BUILD_IN_SOURCE 1
-        INSTALL_COMMAND make install
-      )
+      set( HDF5_TEST_CPP OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_EXAMPLES OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_FORTRAN OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_JAVA OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_PARALLEL OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_PASSTHROUGH_VOL OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_SERIAL OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_SWMR OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_TOOLS OFF CACHE BOOL "" FORCE)
+      set( HDF5_TEST_VFD OFF CACHE BOOL "" FORCE)
+      register_project(hdf5 HDF5 https://github.com/HDFGroup/hdf5 hdf5-1_12_2)
+      unset(CMAKE_RUNTIME_OUTPUT_DIRECTORY CACHE)
+      #Mark CATCH variables as advanced.
+      get_cmake_property(_variableNames VARIABLES)
+      foreach (_variableName ${_variableNames})
+	string(FIND "${_variableName}" "HDF5_" out)
+        if("${out}" EQUAL 0)
+          mark_as_advanced(${_variableName})
+        endif()
+      endforeach()
       set (build_hdf5 TRUE PARENT_SCOPE)
+      set (hdf5_include ${HDF5_INCLUDE_DIRS})
+      set (hdf5_lib ${HDF5_LIBRARIES})
     else ()
       message (STATUS "using contrib HDF5 found at ${hdf5_search}")
     endif ()
     # either it was already here, or we just built it here
-    set (hdf5_include ${hdf5_contrib_path}/include)
-    set (hdf5_lib "-L${hdf5_contrib_path}/lib -Wl,-rpath,${hdf5_contrib_path}/lib/ -lhdf5")
-  endif ()
+    target_include_directories (hdf5 INTERFACE ${hdf5_include})
+    target_link_libraries (hdf5 INTERFACE ${hdf5_lib})
 
-  target_include_directories (hdf5 INTERFACE ${hdf5_include})
-  target_link_libraries (hdf5 INTERFACE ${hdf5_lib})
+  endif ()
 endfunction()
 
 ###############################################################################

--- a/contrib/io.cmake
+++ b/contrib/io.cmake
@@ -39,7 +39,7 @@ function (get_hdf5)
       message (STATUS "libhdf5 not found. We'll build it from source.")
 
       include (ExternalProject)
-      if (DEFINED CMAKE_APPLE_SILICON_PROCESSOR and CMAKE_APPLE_SILICON_PROCESSOR STREQUAL "arm64")
+      if (DEFINED CMAKE_APPLE_SILICON_PROCESSOR AND CMAKE_APPLE_SILICON_PROCESSOR STREQUAL "arm64")
         # Get HDF5 to build on Apple silicon
         ExternalProject_Add (hdf5-ext
           UPDATE_COMMAND ""
@@ -53,7 +53,7 @@ function (get_hdf5)
           INSTALL_COMMAND make install
         )
       else()
-	ExternalProject_Add (hdf5-ext
+        ExternalProject_Add (hdf5-ext
           UPDATE_COMMAND ""
           PREFIX contrib/hdf5
           URL https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.bz2

